### PR TITLE
Adding functional tests for all training and evaluation tasks

### LINKED ISSUE #2621: Create functional tests for all V1 Evaluation scenarios
As laid out in #2498 , we need scenarios to cover the Evaluation functionality we want fully supported in V1.

* I can evaluate a model trained for any of my tasks on test data. The evaluation outputs metrics that are relevant to the task (e.g. AUC, accuracy, P/R, and F1 for binary classification)
* I can get the data that will allow me to plot PR curves

### PR DESCRIPTION
**Pull request by rogancarr**
This PR adds functional tests for all training and evaluation tasks.

* I can evaluate a model trained for any of my tasks on test data. The evaluation outputs metrics that are relevant to the task (e.g. AUC, accuracy, P/R, and F1 for binary classification)
* I can get the data that will allow me to plot PR curves (blocked by #2645)

Fixes #2621

### CODE REVIEW

File: test/Microsoft.ML.Functional.Tests/Evaluation.cs [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+                hasHeader: TestDatasets.mnistOneClass.fileHasHeader,
+                separatorChar: TestDatasets.mnistOneClass.fileSeparator);
+
+            // Create a training pipeline.
+            var pipeline = mlContext.AnomalyDetection.Trainers.RandomizedPca();
+
+            // Train the model.
+            var model = pipeline.Fit(trainData);
+
+            // Evaulate the model.
```
Comment (artidoro) [inline test/Microsoft.ML.Functional.Tests/Evaluation.cs]: >Evaulate [](start = 15, length = 8)

I think there is small typo that was copy pasted everywhere, can you fix that?

File: test/Microsoft.ML.Functional.Tests/Common.cs [hunk 2]
```diff
@@ -160,6 +161,76 @@ public static void AssertEqual(TypeTestData testType1, TypeTestData testType2)
             Assert.True(testType1.Ug.Equals(testType2.Ug));
         }
 
+        /// <summary>
+        /// Check that a <see cref="AnomalyDetectionMetrics"/> object is valid.
+        /// </summary>
+        /// <remarks>
+        /// TODO #2644: At times, AnomalyDetection.Evaluate will return a set of NaN metrics.
+        /// </remarks>
```
Comment (artidoro) [inline test/Microsoft.ML.Functional.Tests/Common.cs]: Maybe would not add the TODO to a xml comment. I would suggest having it as a regular comment only.

File: test/Microsoft.ML.Functional.Tests/Common.cs [hunk 3]
```diff
@@ -160,6 +161,73 @@ public static void AssertEqual(TypeTestData testType1, TypeTestData testType2)
...
+            // Perform sanity checks on the metrics.
+            Assert.InRange(metrics.Auc, 0, 1);
+            Assert.InRange(metrics.DrAtK, 0, 1);
+        }
+
+        /// <summary>
+        /// Check that a <see cref="BinaryClassificationMetrics"/> object is valid.
+        /// </summary>
+        /// <param name="metrics">The metrics object.</param>
+        public static void CheckMetrics(BinaryClassificationMetrics metrics)
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Functional.Tests/Common.cs]: >BinaryClassificationMetrics [](start = 40, length = 27)

we added `CalibratedBinaryClassificationMetrics` recently. Strangely you don't have them in your PR.

Comment (rogancarr) [reply test/Microsoft.ML.Functional.Tests/Common.cs]: Good catch. I hadn't noticed that.

---
In reply to: [259454652](https://github.com/dotnet/machinelearning/pull/2646#discussion_r259454652) [](ancestors = 259454652)

**Comment by rogancarr** [comment]
> Once you fix the build and add the check and related test with `CalibratedBinaryClassificationMetrics` it should be good to go.

This is a weird build error. When the build systems do it, it's not handling the method overloading correctly. I can't repro it locally....

File: test/Microsoft.ML.Functional.Tests/Evaluation.cs [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+            // Train the model.
+            var model = pipeline.Fit(data);
+
+            // Evaluate the model.
+            var scoredData = model.Transform(data);
+            var metrics = mlContext.BinaryClassification.Evaluate(scoredData);
+
+            // Check that the metrics returned are valid.
+            Common.CheckMetrics(metrics);
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Functional.Tests/Evaluation.cs]: >CheckMetrics [](start = 19, length = 12)

So the reason why I mentioned `CalibratedBinaryClassificationMetrics` is not only to add them to Common.cs, but also to have test on them.
Maybe have two pipelines in this test. One with calibrated, one without.

File: test/Microsoft.ML.Functional.Tests/Datasets/Iris.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            // Load the Iris data.
+            var data = mlContext.Data.ReadFromTextFile<Iris>(filePath, hasHeader: hasHeader, separatorChar: separatorChar);
+
+            // Create a function that generates a random groupId.
+            var rng = new Random(seed);
+            Action<Iris, IrisWithGroup> generateGroupId = (input, output) =>
+            {
+                output.Label = input.Label;
+                // The standard set used in tests has 150 rows
+                output.GroupId = (ushort)rng.Next(0, 30);
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Functional.Tests/Datasets/Iris.cs]: >ushort [](start = 34, length = 6)

any reason why it's ushort?

File: test/Microsoft.ML.Functional.Tests/Datasets/TrivialMatrixFactorization.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+
+using System;
+using Microsoft.Data.DataView;
+using Microsoft.ML.Data;
+
+namespace Microsoft.ML.Functional.Tests.Datasets
+{
+    /// <summary>
+    /// A class containing one property per <see cref="DataKind"/>.
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Functional.Tests/Datasets/TrivialMatrixFactorization.cs]: >/// A class containing one property per <see cref="DataKind"/>. [](start = 4, length = 63)

I don't get this sentence. What DataKind has to do with MF? Your data is row+col+value triplets for matrix.

File: test/Microsoft.ML.Functional.Tests/Evaluation.cs (hunk 4 above)
Comment (rogancarr) [reply test/Microsoft.ML.Functional.Tests/Evaluation.cs]: Gotcha ;)

---
In reply to: [259509132](https://github.com/dotnet/machinelearning/pull/2646#discussion_r259509132) [](ancestors = 259509132)

File: test/Microsoft.ML.Functional.Tests/Datasets/TrivialMatrixFactorization.cs (hunk 6 above)
Comment (rogancarr) [reply test/Microsoft.ML.Functional.Tests/Datasets/TrivialMatrixFactorization.cs]: Copy/Paste error, unfortunately.

---
In reply to: [259511254](https://github.com/dotnet/machinelearning/pull/2646#discussion_r259511254) [](ancestors = 259511254)

File: test/Microsoft.ML.Functional.Tests/Datasets/Iris.cs (hunk 5 above)
Comment (rogancarr) [reply test/Microsoft.ML.Functional.Tests/Datasets/Iris.cs]: Remnants of when I was trying to read directly as a `KeyType`. (see #2642)

---
In reply to: [259509894](https://github.com/dotnet/machinelearning/pull/2646#discussion_r259509894) [](ancestors = 259509894)

**Pull request merged**